### ISSUE TITLE
attempted early 2016 with Jupyter

### ISSUE THREAD

**Issue by fomightez** [stuaxo/ipython_cairo#4]
I couldn't quite get your code to work in Jupyter notebook running at mybinder.org in April 2016. It said something about the extension approach you had being deprecated.

So I somewhat adapted your goal and that of [vpython-jupyter](https://github.com/BruceSherwood/vpython-jupyter) to come up with a  a less slick version that works as a Jupyter Binder hosted by mybinder.org, check it out [here](http://mybinder.org/repo/fomightez/cairo-jupyter). Click on one of the examples listed; the link to the image where appear broken but documentation in the examples will tell you what to do to run the code the first time. The github repo it is sourced from can be found [here](https://github.com/fomightez/cairo-jupyter).

**Comment by stuaxo** [comment]
Cheers, I didn't realise this the extension approach was deprecated in jupyter, or heard of mybinder before.
I couldn't get the examples to generate any images, but maybe I need to read up more on mybinder ?

**Comment by fomightez** [comment]
You need to make the associated `.png` file first, and so you need to run all the cells and then reload your browser window. If you enter `ls` in a cell and don't see any files ending in `.png` then you won't have an image to show in the cell yet. 

Easiest way to generate the `.png` image file is to select `Run all` under `Cell` menu and then `Restart` under the `Kernel` menu. (Or, you may be able to do that in one step in some versions of Jupyter by select `Restart & Run all` under the `Kernel` menu.) Then you reload your browser page using your browser's reload feature. You should see an image then.

Later I may just put the image file there in the `demos` folder so an image is already there when the Jupyter notebook loads at http://mybdinder.org/repo/fomightez.cairo-jupyter, but I really wanted to highlight it is an active notebook for now and make user generate it the first time. 

I just added another example, `hello_example.ipynb`, to the list.

More about mybinder.org  can be found at [mybinder.org](http://mybinder.org).

**Comment by stuaxo** [comment]
Managed to get this to render in the end - I had to double click where the image was (the cell with just "title" in it)  then run that.

I'll try and have a look at the code in the next few days.    

The fractal tree is nice, if we could get this working it would be great to have a collection of notebooks showing off cairo, do you know any other nice examples ?

**Comment by fomightez** [comment]
One I thought was nice and planned to convert next was 
http://www.pritschet.me/wiki/python/example-scripts-python/2d-graphics-cairo/

Some others I have seen that could have useful examples to mine are:
http://www.tortall.net/mu/wiki/CairoTutorial

http://zetcode.com/gfx/pycairo/

http://zetcode.com/gfx/cairo/

**Comment by stuaxo** [comment]
I hadn't seen the top one, it looks really good, the other 3 I've used and are very nice.

Originally I was planning to add stuff like being able to introspect cairo paths (e.g. mouse over and see coordinates, maybe have an option to toggle a scaling grid) - would these sorts of features be useful (assuming I can port the 'slick' stuff to Jupyter)  ?

**Comment by fomightez** [comment]
I think your original thinking was spot on. Those additions would be super useful for developing/learning with cairo in the notebook.

I think if I understand correctly, these might be useful as inspiration or coding aids:
http://mpld3.github.io/

http://blog.thedataincubator.com/2015/08/embedding-d3-in-an-ipython-notebook/

http://multithreaded.stitchfix.com/blog/2015/12/15/d3-jupyter/

https://github.com/carnby/datagramas

**Comment by stuaxo** [comment]
Looks like "install_ext" and "uninstall_ext" is deprecated but everything else should be fine if I convert to a python module, will look into this in the next few days.

We should have one notebook ob using ipython_cairo and then just a bunch on different cairo things I reckon.

**Comment by fomightez** [comment]
That is good news; I was hoping it wouldn't be a monumental hurdle to get it working in the newer Jupyter notebooks. 

Given [vpython-jupyter](https://github.com/BruceSherwood/vpython-jupyter) works as a Jupyter Binder hosted by mybinder.org ([here](http://mybinder.org/repo/BruceSherwood/vpython-jupyter)), would you think it would be possible to get your module to work at mybinder.org? Curiously, I did notice the menu options for [vpython-jupyter's](http://mybinder.org/repo/BruceSherwood/vpython-jupyter) notebooks are different for notebooks I've been launching from there, for example look under the `Kernel` menu of a vpython-jupyter notebook, and so I wonder if they had to use a different or newer Jupyter notebook version? It is a fairly recent undertaking and so they might be helpful if you hit a wall on your own.

Your idea about the notebooks seems good. I can certainly help with that effort further. 

Is this time to consider a new repo/different name for the module?

**Comment by stuaxo** [comment]
I'm happy to rename it + also give you commit access to whatever it is / becomes.

I'm a bit torn as to whether it should have ipython or jupyter in the name-  jupyter is the new thing - but isn't just about python, this is about cairo and python, so possibly keeping ipython in the name makes sense ?

I think vpython can use rapydscript and webgl, so those notebooks may not be 'pure' python.

I'm fairly time-poor at the moment, so while I can do things like sort it out as a module, accept patches, and generally help - it probably will take me a lot longer to make some of the more advanced features (though I have shedloads of ideas here).

Having said that I can have a bit of a look at mybinder, but definitely can't promise anything.

**Comment by fomightez** [comment]
I just thought I'd ask about the name while you were considering refactoring. I think you could definitely stick with this one.

Certainly, don't feel rushed at all. The fact that you are interested in maybe getting it working in Jupyter is more than I hoped for.

Plus, I think your time would be better spent on the basics and the advanced features you have planned, and so you can forget about the specialized mybinder.org for now. I have worked out what I was hoping to do in my workaround, at least for basics.

Probably, just submitting pull requests when you get to the notebook development part would work well enough.

**Comment by stuaxo** [comment]
I had a bit of a play with mybinder and it looks fairly straightforward to get things working again.

The main change is that the extension is loaded by putting it in a ```requirements.txt``` - binder will look for one in a folder called ```binder```.   This way we can have the notebooks and the jupyter extension in the same repo.

For debugging pycairo it's great to be able to just put the context or surface and have it displayed, but for the examples I think this looks a bit weird, so I want to see if it will work like

```
from IPython.display import Image
Image(some_context)
```

I think this may "just work" with the extension loaded already.

**Comment by fomightez** [comment]
A-Ha. Progress.  
Two things worked to fix it so I can use cairo in notebooks launched from MyBinder.org. My better understanding of using requirements to install cairocffi and how to import it within the notebook, and your suggestion to use the following:

`from IPython.display import Image
Image(some_context)`

If you go to [here](https://github.com/fomightez/cairo-jupyter) and use the launch button, all the examples work now. Yay.

I guess this is a lot more direct in a way and hardly much more code, really. Thanks so much for helping me!

**Comment by stuaxo** [comment]
Now I understand you binder more, I appreciate the simplicity of the approach you have.

Having said that, it would be great if the extension could live in the same repo, if I can get it working, are you up for it living in the jupyter-cairo repo - since I have way less time than even when I made the original extension.

**Comment by fomightez** [comment]
Definitely. I'd be happy to help host the updated version of the extension wherever you'd like.

**Comment by stuaxo** [comment]
Got something sort of working   https://github.com/stuaxo/cairo-jupyter    I'll try and make a proper pull request, in the next week or two on your repo :)

**Comment by fomightez** [comment]
That looks great.  
My only question would be whether that `%reload_ext cairo_jupyter` command is necessary? They seem to have them loaded already at https://github.com/binder-examples/jupyter-extension . Can you add something like `jupyter nbextension enable --py  cairo_jupyter` or `jupyter nbextension enable cairo_jupyter` to the postBuild file?

**Comment by stuaxo** [comment]
I couldn't get it working locally or on mybinder when I tried it last night.
(sorry can't paste it here as I'm at work). 


I'm wondering if it gets confused - the postBuild commands run in the root, so maybe enabling cairo_jupyter finds the "cairo_jupyter" folder in the directory it is in, instead of the installed module.

I could probably try tweaking setup.py and move the module some where like ```nbextension/cairo_jupyter``` and see if that helps.


Probably won't have any time to look at this tonight.

Not needing %reload would definitely be better.

**Comment by fomightez** [comment]
No rush to look into it at all. Just was curious about it. Even with needing that line, it is super nice to have the extension working with the BInderized-version.
